### PR TITLE
Update handling of response

### PR DESCRIPTION
**Pull request by tomazpu**

### CODE REVIEW

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Coverage increased per file
===========================
+ wirecardpaymentgateway/classes/Response/Success.php  7
         

Coverage decreased per file
===========================
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  -6
         

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 1]
```diff
@@ -86,63 +85,49 @@ private function determineFinalOrderId()
...
     {
+        $transactionService = new TransactionService($config, new WirecardLogger());
+
         $isSeamlessTransaction = \Tools::getValue('jsresponse');
+
+        //@TODO refactor to a known design pattern.
         if ($isSeamlessTransaction) {
-            return $this->executeSeamlessTransaction($_POST, $config, $cart, $orderId);
+            $response = $this->executeSeamlessTransaction($_POST, $transactionService);
+        } else {
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [The method determineFinalOrderId uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4518447)

File: wirecardpaymentgateway/classes/Response/PostProcessing/Success.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        parent::process();
+
+        $transaction_id = \Tools::getValue('tx_id');
+
+        $this->transaction_manager->markTransactionClosed($transaction_id);
+        $this->context_service->setConfirmations(
+            $this->getTranslatedString('success_new_transaction')
+        );
+    }
+}
```
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Response/PostProcessing/Success.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4518447)

File: wirecardpaymentgateway/classes/Response/Initial/Success.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+        \Tools::redirect(
+            'index.php?controller=order-confirmation&id_cart='
+            .$this->cart->id.'&id_module='
+            .$this->module->id.'&id_order='
+            .$this->order->id.'&key='
+            .$this->customer->secure_key
+        );
+    }
+}
```
Comment (rinnhofer) [inline wirecardpaymentgateway/classes/Response/Initial/Success.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4518447)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -2
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -2
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/Service/OrderService.php  3
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Config/SepaConfigurationFactory.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  4
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -2
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         

Clones added
============
- wirecardpaymentgateway/helper/Service/OrderService.php  1
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  1
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  4
- wirecardpaymentgateway/helper/Service/OrderService.php  3
- wirecardpaymentgateway/classes/Config/SepaConfigurationFactory.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -2
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         

Clones added
============
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  1
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  1
- wirecardpaymentgateway/helper/Service/OrderService.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  4
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Config/SepaConfigurationFactory.php  1
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/helper/Service/OrderService.php  3
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         

Clones added
============
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  1
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  1
- wirecardpaymentgateway/helper/Service/OrderService.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  2
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  4
- wirecardpaymentgateway/helper/Service/ContextService.php  2
- wirecardpaymentgateway/helper/Service/OrderService.php  3
- wirecardpaymentgateway/classes/Config/GenericConfigurationFactory.php  2
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  4
- wirecardpaymentgateway/classes/Config/SepaConfigurationFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Success.php  -2
+ wirecardpaymentgateway/controllers/front/payment.php  -1
         

Clones added
============
- wirecardpaymentgateway/helper/Service/ShopConfigurationService.php  1
- wirecardpaymentgateway/classes/Config/CreditcardConfigurationFactory.php  1
- wirecardpaymentgateway/helper/Service/ContextService.php  1
- wirecardpaymentgateway/helper/Service/OrderService.php  1
- wirecardpaymentgateway/classes/Config/PaymentConfigurationFactory.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

File: wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php [hunk 4]
```diff
@@ -356,16 +351,16 @@ private function isValidAddress($module, $shipping, $billing)
             }
         }
 
-        if (count($this->getAllowedCountries($module, 'shipping'))) {
+        if (count($this->getAllowedCountries( 'shipping'))) {
```
Comment (rinnhofer) [inline wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php]: Codacy Issue found: [Space after opening parenthesis of function call prohibited](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4518447)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

File: wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php (hunk 4 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php]: Codacy Issue found: [Space after opening parenthesis of function call prohibited](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4518447)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

File: wirecardpaymentgateway/models/Payment.php [hunk 5]
```diff
@@ -202,12 +202,11 @@ public function setLoadJs($load)
     /**
      * Check if payment is available for specific cart content default true
      *
-     * @param \WirecardPaymentGateway $module
      * @param \Cart $cart
      * @return bool
      * @since 1.0.0
      */
-    public function isAvailable($module, $cart)
+    public function isAvailable($cart)
```
Comment (rinnhofer) [inline wirecardpaymentgateway/models/Payment.php]: Codacy Issue found: [Avoid unused parameters such as '$cart'.](https://app.codacy.com/app/Wirecard/prestashop-ee/pullRequest?prid=4518447)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

File: Test/Unit/Models/Payments/PaymentPaypalTest.php [hunk 6]
```diff
@@ -65,19 +65,13 @@ public function testCreateTransaction()
         $this->assertEquals($expected, $actual::NAME);
     }
 
-    public function testCreateCancelTransaction()
+    public function testGetPostProcessingMandatoryEntities()
```
Comment (ozada) [inline Test/Unit/Models/Payments/PaymentPaypalTest.php]: This method duplicated in all payment methods. What was intention here?  Doesn't it make sense, test it once in abstract class?

File: Test/Unit/WirecardPaymentGatewayTest.php [hunk 7]
```diff
@@ -85,18 +85,6 @@ public function testNotificationUrl()
...
-        $this->assertCount(9, $actual);
-    }
-
-    /*public function testHookActionFrontControllerSetMedia()
-    {
-        $this->assertEquals(true, $this->gateway->hookActionFrontControllerSetMedia());
-    }*/
-
     public function testExecuteSql()
     {
```
Comment (ozada) [inline Test/Unit/WirecardPaymentGatewayTest.php]: What's covered here?

File: wirecardpaymentgateway/classes/Response/Initial/Success.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    /**
+     * Check if the payment method is PoiPia and the payment type is Pia
+     *
+     * @return bool
+     * @since 2.5.0
+     */
+    private function isPiaPayment()
+    {
```
Comment (ozada) [inline wirecardpaymentgateway/classes/Response/Initial/Success.php]: I think it should be extracted to appropriate nested class.

File: wirecardpaymentgateway/classes/Response/PostProcessing/Success.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    private $context_service;
+
+    /**
+     * Success constructor.
+     * @param $order
+     * @param $response
+     * @since 2.5.0
+     */
+    public function __construct($order, $response)
```
Comment (ozada) [inline wirecardpaymentgateway/classes/Response/PostProcessing/Success.php]: Args DI is possible since 5.0.x

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/ProcessablePaymentResponseFactory.php  1
- wirecardpaymentgateway/classes/Response/Initial/Success.php  2
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  1
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/controllers/front/payment.php  -1
+ wirecardpaymentgateway/classes/Response/Success.php  -2
+ wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/prestashop-ee/prestashop-ee/pullRequest?prid=4518447&bid=15220781)

File: wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php [hunk 10]
```diff
@@ -438,4 +425,39 @@ public function getPostProcessingMandatoryEntities()
...
         ];
     }
+
+    /**
+     * @return \Cart
+     * @since 2.5.0
+     */
+    protected function getCartFromContext()
+    {
+        $context = \Context::getContext();
```
Comment (ozada) [inline wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php]: exception if cart is null?

File: Test/Unit/Models/Payments/PaymentPaypalTest.php (hunk 6 above)
Comment (tomazpu) [reply Test/Unit/Models/Payments/PaymentPaypalTest.php]: If i test it once in the abstract there is no guarantee that somebody didn't add an entity by mistake to a payment method.

File: Test/Unit/WirecardPaymentGatewayTest.php (hunk 7 above)
Comment (tomazpu) [reply Test/Unit/WirecardPaymentGatewayTest.php]: No idea not my code?

File: wirecardpaymentgateway/classes/Response/Initial/Success.php (hunk 8 above)
Comment (tomazpu) [reply wirecardpaymentgateway/classes/Response/Initial/Success.php]: Sure but not at this stage

File: wirecardpaymentgateway/classes/Response/PostProcessing/Success.php (hunk 9 above)
Comment (tomazpu) [reply wirecardpaymentgateway/classes/Response/PostProcessing/Success.php]: Not in any class that is written. After the coding standards im the first for it.

File: wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php (hunk 10 above)
Comment (tomazpu) [reply wirecardpaymentgateway/models/PaymentGuaranteedInvoiceRatepay.php]: Definition of presta is that always a cart is returned so checking for null makes no sense in my opinion.

**Pull request merged**